contracts/FlowIDTableStaking.cdc

```diff
@@ -870,6 +870,79 @@ pub contract FlowIDTableStaking {
             }
         }
 
+        /// Set slot (count) limits for each node type (role).
+        pub fun setSlotLimits(slotLimits: {UInt8: UInt8}) {
+            FlowIDTableStaking.account.load<{UInt8: UInt8}>(from: /storage/flowStakingSlotLimits)
+            FlowIDTableStaking.account.save(slotLimits, to: /storage/flowStakingSlotLimits)
+        }
+
+        /// Enforces slot (count) limits for each node type (role). If more nodes
+        /// of a given type are registered than the limit, they are removed
+        /// randomly until the limits are met. When a node is removed, it's
+        /// committed tokens are transfered to it's unstaked bucket.
+        pub fun slotSelect() {
+            let allNodeIDs = FlowIDTableStaking.getNodeIDs()
+
+            let slotLimits = FlowIDTableStaking.account.load<{UInt8: UInt8}>(from: /storage/flowStakingSlotLimits)
+                ?? panic("Could not load the slot limits object from storage!")
+
+            // count the nodes of each type
+            var nodeCounts: {UInt8: UInt8} = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+            for nodeID in allNodeIDs {
+                let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
+                if nodeRecord.initialWeight == 100 {
+                    nodeCounts[nodeRecord.role] = nodeCounts[nodeRecord.role]! + 1
+                }
+            }
+
+            for key in nodeCounts.keys {
+                while nodeCounts[key]! > slotLimits[key]! {
```

Comment:
We will need a way to omit some node roles from slot selection entirely, because we cannot immediately allow public access to all node roles.